src/dlfcn.c

```diff
@@ -467,7 +472,161 @@ char *dlerror( void )
+     * 00401002  |. E8 7B0D0000    CALL 00401D82               ; \GetModuleHandleA
+     * ...thunk table...
+     * 00401D82   $-FF25 4C204000 , JMP DWORD PTR DS:[40204C]  ;  KERNEL32.GetModuleHandleA
+     * ...memory address value of pointer...
+     * 40204C > FC 3D 57 7C   ;little endian pointer value
+     */
+    if( *(short *)addr != 0x25ff )
+        return NULL;
+
+    /* get offset from thunk table (after instruction 0xff 0x25)
+     *   4018c8 <_VirtualQuery>: ff 25 4a 8a 00 00
+     */
+    ULONG offset = *(ULONG*)(addr+2);
+#ifdef _WIN64
+    /* On 64 bit the offset is relative
+     *   4018c8:	ff 25 4a 8a 00 00    	jmpq   *0x8a4a(%rip)        # 40a318 <__imp_VirtualQuery> # (64bit)
+     */
+    void **ptr = (void *)(addr + 6 + offset);
+#else
+    /* On 32 bit the offset is absolute
+     *    4019b4:	ff 25 90 71 40 00    	jmp    *0x40719
+     */
+    void **ptr = (void *)offset;
+#endif
+    if( ptr < (void**)iat || ptr > (void**)iat + iat_size )
+        return NULL;
+    return *ptr;
+}
+
+/* holds module filename */
+static char _module_filename[2*MAX_PATH];
+
+/**
+ * Get module information (filename and base address) from the address given
+ * @param addr address to get module info for
+ * @param info pointer to store module info
+ * @return TRUE requested info filled into structure pointed by parameter info
+ * @return FALSE error
+ */
+static BOOL getModuleInfo( const void *addr, Dl_info *info )
+{
+    HMODULE hModule;
+    DWORD sLen;
+
+    if (!GetModuleHandleExA( GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, addr, &hModule))
+        return FALSE;
+
+    if( !hModule )
+    {
+        return FALSE;
+    }
+
+    info->dli_fbase = (void *)hModule;
+
+    sLen = GetModuleFileNameA( hModule, _module_filename, sizeof( _module_filename ) );
+    if( sLen == 0 )
+        return FALSE;
+    if( sLen == sizeof( _module_filename ) && GetLastError() == ERROR_INSUFFICIENT_BUFFER )
+        return FALSE;
+    info->dli_fname = _module_filename;
+    return TRUE;
+}
+
+DLFCN_EXPORT
+int dladdr( void *addr, Dl_info *info )
+{
+    void *iat_addr = NULL;
+    void *real_addr;
+    HMODULE module;
+    IMAGE_IMPORT_DESCRIPTOR *iat;
+    DWORD size = 0;
+
+    if( !info || !addr )
+        return 0;
+
+    module = GetModuleHandle( 0 );
+    iat = getImportTable( module, &size );
+    if( iat )
+        iat_addr = getAddressFromIAT( iat, size, addr );
+    real_addr = iat_addr ? iat_addr : addr;
+    if( !getModuleInfo( real_addr, info ))
+    {
+        info->dli_fname = NULL;
+        info->dli_fbase = NULL;
+        info->dli_saddr = NULL;
+    }
+    else
+    {
+        info->dli_saddr = (void*)real_addr;
+        if (iat) {
+            char *sym = getSymbolName( module, iat, real_addr );
+            if (sym) {
+                info->dli_sname = sym;
+                return 1;
+            }
+        }
+    }
+    info->dli_sname = NULL;
+    return 1;
+}

```

Comment:
Suggestion how to simplify code to not have too many open code blocks for if-conditions.
```c
static char module_filename[2*MAX_PATH];

static void fillModuleInfo( void *addr, Dl_info *info )
{
    HMODULE hModule;
    DWORD dwSize;

    if( !GetModuleHandleExA( GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, addr, &hModule ) || !hModule )
    {
        info->dli_fname = NULL;
        info->dli_fbase = NULL;
        return;
    }

    info->dli_fbase = (void *) hModule;

    dwSize = GetModuleFileNameA( hModule, module_filename, sizeof( _module_filename ) );
    if( dwSize > 0 && dwSize < sizeof( module_filename ) )
        info->dli_fname = module_filename;
    else
        info->dli_fname = NULL;
}

DLFCN_EXPORT
int dladdr( void *addr, Dl_info *info )
{
    void *iatAddr;
    void *realAddr;
    HMODULE hModule;
    IMAGE_IMPORT_DESCRIPTOR *iat;
    DWORD dwSize;

    if( !addr || !info )
        return 0;

    hModule = GetModuleHandleA( NULL );
    if( !hMoudule )
        return 0;

    iat = getImportTable( hModule, &dwSize );
    iatAddr = iat ? getAddressFromIAT( iat, dwSize, addr ) : NULL;
    realAddr = iatAddr ? iatAddr : addr;

    fillModuleInfo( realAddr, info );

    if( !iat && !info->dli_fbase )
        return 0;

    info->dli_sname = iat ? getSymbolName( hModule, iat, realAddr ) : NULL;
    info->dli_saddr = (void *) realAddr;

    return 1;
}
```